src/main/scala/viper/gobra/frontend/info/implementation/typing/ExprTyping.scala

```diff
@@ -333,13 +342,60 @@ trait ExprTyping extends BaseTyping { this: TypeInfoImpl =>
+              val lIsInteger = assignableTo.errors(l, UNTYPED_INT_CONST)(n)
+              val rIsInteger = assignableTo.errors(r, UNTYPED_INT_CONST)(n)
+              val typesAreMergeable = mergeableTypes.errors(l, r)(n)
+              val exprWithinBounds = {
+                if(typesAreMergeable.isEmpty) {
+                  // Only makes sense to check that a binary expression is within bounds if the types of its
+                  // subexpressions can be combined
+
+                  // mergedType must exist because, otherwise typesAreMergeable.isEmpty would not hold
+                  val mergedType = typeMerge(l, r).get
+
+                  // The first two checks ensure that, if an operand is constant, then it must be assignable to the type
+                  // of the result. This makes the type system capable of rejecting expressions like `uint8(1) * (-1)`,
+                  // which are also rejected by the go compiler
+                  intExprWithinTypeBounds(n.left.asInstanceOf[PExpression], mergedType) ++
+                    intExprWithinTypeBounds(n.right.asInstanceOf[PExpression], mergedType) ++
+                    intExprWithinTypeBounds(n, mergedType)
+                } else noMessages
+              }
+              lIsInteger ++ rIsInteger ++ typesAreMergeable ++ exprWithinBounds
             }
+          case (_: PShiftLeft, l, r) =>
+            val integerOperands = assignableTo.errors(l, UNTYPED_INT_CONST)(n) ++ assignableTo.errors(r, UNTYPED_INT_CONST)(n)
+            if (integerOperands.isEmpty) {
+              intConstantEval(n.right.asInstanceOf[PExpression]) match {
+                case Some(v) =>
+                  // The Go compiler checks that the RHS of (<<) is non-negative and, at most, the size
+                  // of the type of the left operand (or 512 if there is an untyped const on the left)
+                  val lowerBound = error(n.right, s"constant ${n.right} overflows uint", v < 0)
+                  val nBits = exprOrTypeType(n.left) match {
+                    case IntT(t: BoundedIntegerKind) => t.nbits
+                    case IntT(UnboundedInteger) => MAX_SHIFT
+                    case t => violation(s"unexpected type $t")
+                  }
+                  val upperBound = error(n.right, s"shift count ${n.right} large for type ${exprOrTypeType(n.left)}", v > nBits)
+                  lowerBound ++ upperBound
+
+                case None => noMessages
+              }
+            } else integerOperands
+          case (_: PShiftRight, l, r) =>
+            val integerOperands = assignableTo.errors(l, UNTYPED_INT_CONST)(n) ++ assignableTo.errors(r, UNTYPED_INT_CONST)(n)
+            if (integerOperands.isEmpty) {
+              (intConstantEval(n.right.asInstanceOf[PExpression]) match {
+                case Some(v) =>
+                  // The Go compiler only checks that the RHS of (>>) is non-negative
+                  error(n, s"constant $r overflows uint", v < 0)
```

Comment:
same here, I would simply create an error message stating that the constant is negative